ui/gtk4.c

```diff
@@ -0,0 +1,2176 @@
+                          GtkDisplayState *s)
+{
+    if (s->external_pause_update) {
+        return;
+    }
+    if (runstate_is_running()) {
+        qmp_stop(NULL);
+    } else {
+        qmp_cont(NULL);
+    }
+}
+
+static void gd_menu_reset(void *opaque)
+{
+    qmp_system_reset(NULL);
+}
+
+static void gd_menu_powerdown(void *opaque)
+{
+    qmp_system_powerdown(NULL);
+}
+
+static void gd_menu_quit(GSimpleAction *action, GVariant *param,
+                         GtkDisplayState *s)
+{
+    qmp_quit(NULL);
+    g_application_quit(G_APPLICATION(s->app));
+}
+
+static void gd_menu_switch_vc(GSimpleAction *action, GVariant *param,
+                              GtkDisplayState *s)
+{
+    VirtualConsole *vc = gd_vc_find_by_menu(s,
+                                            g_variant_get_string(param, NULL));
+    GtkNotebook *nb = GTK_NOTEBOOK(s->notebook);
+    gint page;
+
+    gtk_release_modifiers(s);
+    if (vc) {
+        page = gtk_notebook_page_num(nb, vc->tab_item);
+        gtk_notebook_set_current_page(nb, page);
+        g_simple_action_set_state(action, param);
+        gtk_widget_grab_focus(vc->focus);
+    }
+}
+
+static void gd_menu_show_tabs(GSimpleAction *action, GVariant *param,
+                              GtkDisplayState *s)
+{
+    VirtualConsole *vc = gd_vc_find_current(s);
+
+    if (!s_action_get_state(s->show_tabs_action)) {
+        gtk_notebook_set_show_tabs(GTK_NOTEBOOK(s->notebook), TRUE);
+        s_action_set_state(s->show_tabs_action, TRUE);
+    } else {
+        gtk_notebook_set_show_tabs(GTK_NOTEBOOK(s->notebook), FALSE);
+        s_action_set_state(s->show_tabs_action, FALSE);
+    }
+    gd_update_windowsize(vc);
+}
+
+static gboolean gd_tab_window_close(GtkWindow *widget, VirtualConsole *vc)
+{
+    GtkDisplayState *s = vc->s;
+
+    /*
+     * FIXME: mark the menu item as sensitive again
+     * gtk_widget_set_sensitive(vc->menu_item, true);
+     */
+    g_object_ref(vc->tab_item);
+    gtk_window_set_child(GTK_WINDOW(vc->window), NULL);
+    gtk_notebook_append_page(GTK_NOTEBOOK(s->notebook), vc->tab_item, NULL);
+    g_object_unref(vc->tab_item);
+    gtk_notebook_set_tab_label_text(GTK_NOTEBOOK(s->notebook),
+                                    vc->tab_item, vc->label);
+
+    gtk_window_destroy(GTK_WINDOW(vc->window));
+    vc->window = NULL;
+    if (vc->gfx.esurface) {
+        eglDestroySurface(qemu_egl_display, vc->gfx.esurface);
+        vc->gfx.esurface = NULL;
+    }
+    if (vc->gfx.ectx) {
+        eglDestroyContext(qemu_egl_display, vc->gfx.ectx);
+        vc->gfx.ectx = NULL;
+    }
+    return TRUE;
+}
+
+static gboolean gd_win_grab(GtkWidget *widget, GVariant *args,
+                            void *opaque)
+{
+    VirtualConsole *vc = opaque;
+    fprintf(stderr, "%s: %s\n", __func__, vc->label);
+    if (vc->s->ptr_owner) {
+        gd_ungrab_pointer(vc->s);
+    } else {
+        gd_grab_pointer(vc, "user-request-detached-tab");
+    }
+    return TRUE;
```

Comment:
Do you need the return value?

Reply:
It actually doesn't, good point. Thanks!

Reply:
Actually it is needed and true means the action was handled successfully

Reply:
There is no branch where it would return `FALSE` though, no?

Reply:
yes, the action never fails in our case, I documented why the return is needed